File: warehouse/go/product.go
package main

import (
	"fmt"
	"math/rand"
	"strings"
	"time"
)

// Глобальные переменные для склада
var warehouse = make(map[string]map[string]int)
var racks = []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
var sections = []int{1, 2, 3, 4, 5, 6, 7}
var shelves = []int{1, 2, 3, 4}

const MAX_IN_CELL = 10

// Инициализация склада (создание всех ячеек)
func initWarehouse() {
	zone := "A"
	for _, r := range racks {
		for _, s := range sections {
			for _, sh := range shelves {
				address := fmt.Sprintf("%s%02d%d%d", zone, r, s, sh)
				warehouse[address] = make(map[string]int)
			}
		}
	}
}

// Добавление товара в ячейку склада
func addItem(item, address string, count int) {
	cell, ok := warehouse[address]
	if !ok {
		fmt.Printf("Ячейка %s не найдена\n", address)
		return
	}
	total := 0
	for _, quant := range cell {
		total += quant
	}
	if total+count > MAX_IN_CELL {
		fmt.Printf("Нет места в ячейке %s\n", address)
		return
	}
	cell[item] += count
	fmt.Printf("Добавлен %s в ячейку %s\n", item, address)
}

// Удаление товара из ячейки склада
func removeItem(item, address string, count int) {
	cell, ok := warehouse[address]
	if !ok {
		fmt.Printf("Ячейка %s не найдена\n", address)
		return
	}
	quant, ok := cell[item]
	if !ok {
		fmt.Printf("Товар %s не найден\n", item)
		return
	}
	if count <= quant {
		cell[item] -= count
		if cell[item] == 0 {
			delete(cell, item)
		}
		fmt.Printf("Товар %s удален из ячейки %s в количестве %d\n", item, address, count)
	} else {
		fmt.Printf("Число %d больше, чем количество %s\n", count, item)
	}
}

// Информация о складе: загрузка, зоны, содержимое и пустые ячейки
func info() {
	totalCells := len(warehouse)
	totalCapacity := totalCells * MAX_IN_CELL
	totalItems := 0
	zoneItems := make(map[string]int)
	zoneCells := make(map[string]int)
	emptyCells := []string{}

	for address, cell := range warehouse {
		zone := address[:1]
		cellSum := 0
		for _, count := range cell {
			cellSum += count
		}
		totalItems += cellSum
		zoneItems[zone] += cellSum
		zoneCells[zone]++

		if cellSum == 0 {
			emptyCells = append(emptyCells, address)
		} else {
			fmt.Printf("Ячейка %s: ", address)
			for item, count := range cell {
				fmt.Printf("%s (%d) ", item, count)
			}
			fmt.Println()
		}
	}

	if totalCapacity == 0 {
		fmt.Println("Склад пуст, загрузка: 0%")
	} else {
		warehouseLoad := 100.0 * float64(totalItems) / float64(totalCapacity)
		fmt.Printf("Загрузка склада: %.2f%%\n", warehouseLoad)
		for zone, items := range zoneItems {
			zoneCapacity := zoneCells[zone] * MAX_IN_CELL
			var zoneLoad float64
			if zoneCapacity == 0 {
				zoneLoad = 0
			} else {
				zoneLoad = 100.0 * float64(items) / float64(zoneCapacity)
			}
			fmt.Printf("Загрузка зоны %s: %.2f%%\n", zone, zoneLoad)
		}
	}

	fmt.Println("Пустые ячейки:")
	fmt.Println(strings.Join(emptyCells, " "))
}

func main() {
	rand.Seed(time.Now().UnixNano())
	initWarehouse()

	items := []string{"Молоко", "Хлеб", "Яблоко", "Сыр", "Мясо", "Банан", "Яблоко"}

	// Рандомно заполняем склад
	for address := range warehouse {
		if rand.Intn(2) == 0 {
			continue
		}
		item := items[rand.Intn(len(items))]
		count := 1 + rand.Intn(MAX_IN_CELL)
		addItem(item, address, count)
	}

	// Меню пользователя
	for {
		fmt.Print("\nВыберите действие: ADD, REMOVE, INFO, EXIT\n> ")
		var command string
		fmt.Scan(&command)
		switch command {
		case "ADD":
			var item, address string
			var count int
			fmt.Print("Введите товар, адрес и количество: ")
			fmt.Scan(&item, &address, &count)
			addItem(item, address, count)
		case "REMOVE":
			var item, address string
			var count int
			fmt.Print("Введите товар, адрес и количество: ")
			fmt.Scan(&item, &address, &count)
			removeItem(item, address, count)
		case "INFO":
			info()
		case "EXIT":
			return
		default:
			fmt.Println("Неизвестная команда!")
		}
	}
}

File: warehouse/product.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <iomanip>
#include <cstdlib>
#include <ctime>
using namespace std;

// Глобальные переменные для склада
map<string, map<string, int>> warehouse; // Склад: адрес -> (товар -> количество)
vector<int> racks = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}; // Стеллажи
vector<int> sections = {1, 2, 3, 4, 5, 6, 7}; // Секции
vector<int> shelves = {1, 2, 3, 4}; // Полки
const int MAX_IN_CELL= 10; // Максимум в ячейке

// Инициализация склада (создание всех ячеек)
void initWarehouse() {
    string zone = "A"; // Одна зона

    for (const auto& r : racks) {
        for (const auto& s : sections) {
            for (const auto& sh : shelves) {
                // Формируем адрес: A + номер стеллажа (2 цифры) + секция + полка
                string address = zone + 
                    (r < 10 ? "0" + to_string(r) : to_string(r)) + 
                    to_string(s) + 
                    to_string(sh);

                warehouse[address] = {}; // Инициализируем пустую ячейку
            }
        }
    }
}

// Добавление товара в ячейку склада
void addItem(const string& item, const string& address, int count){
    if(warehouse.find(address) == warehouse.end()){
        cout << "Ячейка" << address << "не найдена" << endl;
        return;
    }

    int total = 0;
    for(const auto& [name, quant] : warehouse[address]){
        total += quant;
    }

    if(total + count > MAX_IN_CELL){
        cout << "Нет места в ячейке" << address << endl;
        return;
    }

    warehouse[address][item] += count;
    cout << "Добавлен " << item << " в ячейку " << address << endl;
}

// Удаление товара из ячейки склада
void removeItem(const string& item, const string& address, int count){
    if(warehouse.find(address) == warehouse.end()){
        cout << "Ячейка " << address << " не найдена" << endl;
        return;
    }

    auto& cell = warehouse[address];
    if(cell.find(item) == cell.end()){
        cout << "Товар " << item << " не найден" << endl;
        return;
    }

    if (count <= cell[item]){
        cell[item] -= count;
        if(cell[item] == 0){
            cell.erase(item);
        }
        cout << "Товар " << item << " удален из ячейки " << address << " в количестве "
             << count << endl;
    } else {
        cout << "Число " << count << " больше, чем количество " << item << endl;
    }
}

// Информация о складе: загрузка, зоны, содержимое и пустые ячейки
void info(){
    int total_cells = warehouse.size();
    int total_capacity = total_cells * MAX_IN_CELL;
    int total_items = 0;
    map<string, int> zone_items;
    map<string, int> zone_cells;

    vector<string> empty_cells;

    for (const auto& [address, cell] : warehouse) {
        string zone = address.substr(0, 1);
        int cell_sum = 0;
        for (const auto& [item, count] : cell) {
            cell_sum += count;
        }
        total_items += cell_sum;
        zone_items[zone] += cell_sum;
        zone_cells[zone]++;

        if (cell_sum == 0) {
            empty_cells.push_back(address);
        } else {
            cout << "Ячейка " << address << ": ";
            for (const auto& [item, count] : cell) {
                cout << item << " (" << count << ") ";
            }
            cout << endl;
        }
    }

    if (total_capacity == 0) {
        cout << "Склад пуст, загрузка: 0%" << endl;
    } else {
        double warehouse_load = (100.0 * total_items / total_capacity);
        cout << fixed << setprecision(2);
        cout << "Загрузка склада: " << warehouse_load << "%" << endl;

        for (const auto& [zone, items] : zone_items) {
            int zone_capacity = zone_cells[zone] * MAX_IN_CELL;
            double zone_load = (zone_capacity == 0) ? 0 : (100.0 * items / zone_capacity);
            cout << "Загрузка зоны " << zone << ": " << zone_load << "%" << endl;
        }
    }

    cout << "Пустые ячейки:" << endl;
    for (const auto& addr : empty_cells) {
        cout << addr << " ";
    }
    cout << endl;
}

// Главная функция: меню пользователя
int main() {
    srand(static_cast<unsigned int>(time(0)));
    initWarehouse();

    vector<string> items = {"Молоко", "Хлеб", "Яблоко", "Сыр", "Мясо", "Банан", "Яблоко"};

    // Рандомно заполняем склад (range-based for)
    for (const auto& [address, _] : warehouse) {
        if (rand() % 2 == 0) continue; // Пропускаем часть ячеек для пустых
        string item = items[rand() % items.size()];
        int count = 1 + rand() % MAX_IN_CELL;
        addItem(item, address, count);
    }

    // Меню пользователя
    while (true) {
        cout << "\nВыберите действие: ADD, REMOVE, INFO, EXIT\n> ";
        string command;
        cin >> command;

        if (command == "ADD") {
            string item, address;
            int count;
            cout << "Введите товар, адрес и количество: ";
            cin >> item >> address >> count;
            addItem(item, address, count);
        } else if (command == "REMOVE") {
            string item, address;
            int count;
            cout << "Введите товар, адрес и количество: ";
            cin >> item >> address >> count;
            removeItem(item, address, count);
        } else if (command == "INFO") {
            info();
        } else if (command == "EXIT") {
            break;
        } else {
            cout << "Неизвестная команда!" << endl;
            continue;
        }
    }

    return 0;
}